### PR TITLE
p2p: Implement a BaseService class and use it everywhere

### PR DESCRIPTION
**Pull request by gsalgado**
We had several classes for our different services and they all had
slightly different APIs for running/stopping them. BaseService aims
to provide a standard API for that.

Includes #673 , so better to review that one first

### CODE REVIEW

File: p2p/service.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        try:
+            await self._run()
+        except OperationCancelled as e:
+            self.logger.info("%s finished: %s", self, e)
+        except Exception:
+            self.logger.exception("Unexpected error in %s, exiting", self)
+        finally:
+            self._finished.set()
+
+    async def stop(self) -> None:
```
Comment (jannikluhn) [inline p2p/service.py]: Should this raise an exception (or log a warning) when trying to stop a service that hasn't been started yet?

File: p2p/service.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        except asyncio.futures.TimeoutError:
+            self.logger.info("Timed out waiting for %s to finish, exiting anyway", self)
+
+    #
+    # These methods should be implemented by subclasses
+    #
+    async def _run(self) -> None:
+        """Runs the service's loop.
+
+        Should return when the CancelToken is triggered.
```
Comment (jannikluhn) [inline p2p/service.py]: I interpret this comment as `_run` should catch `OperationCancelled` exceptions and return cleanly in those cases. But would this be necessary considering that `run` catches them itself and logs a nice message as well?

Comment (gsalgado) [reply p2p/service.py]: The idea is that `_run()` can either return or let `OperationCancelled` bubble up, so I'll amend the docstring to make that clear

File: p2p/service.py (hunk 1 above)
Comment (gsalgado) [reply p2p/service.py]: I don't think that adds any value at this point, and it'd require another event/attribute to track whether or not run() has been called, so I'd rather not do that now

File: p2p/service.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import logging  # noqa: F401
+from typing import Callable, Optional
+
+from p2p.cancel_token import CancelToken
+from p2p.exceptions import OperationCancelled
+
+
+class BaseService(ABC):
+    # Subclasses must define their own logger.
+    logger = None  # type: logging.Logger
```
Comment (pipermerriam) [inline p2p/service.py]: We can use python 3.6 syntax for these types.

```python
logger: logging.Logger = None
```

Comment (pipermerriam) [reply p2p/service.py]: I'm with you but I don't think it's fair to do it anytime in the near term future.  We *just* forced everyone to upgrade to 3.x and while `3.5 -> 3.6` is not a significant as `2.x -> 3.x` it's still not something we can just throw at the community imho.  I'm sure that there are people for whom the upgrade would require some non-trivial amount of infrastructure upgrades to get all of their ops infrastructure in a 3.6 environment.

Comment (cburgdorf) [reply p2p/service.py]: That would also allow ethpm to start with 3.6. But I can't say much about the community backing of such a move.

**Pull request merged**